tools/clang/test/CodeGenHLSL/quick-test/matrix_operators.hlsl

```diff
@@ -0,0 +1,131 @@
+// RUN: %dxc -E main -T vs_6_0 %s | FileCheck %s
+
+// Tests the implementation of unary and binary matrix operators
+
+// Workaround for AppendStructuredBuffer<matrix>.Append bug
+#define Append(buf, val) buf[buf.IncrementCounter()] = (val)
+
+RWStructuredBuffer<int1x1> output_i;
+RWStructuredBuffer<uint1x1> output_u;
+RWStructuredBuffer<float1x1> output_f;
+RWStructuredBuffer<bool1x1> output_b;
+
+void main()
+{
+    int1x1 i1 = int1x1(1);
+    int1x1 i2 = int1x1(2);
+    int1x1 i3 = int1x1(3);
+    int1x1 im1 = int1x1(-1);
+    int1x1 im3 = int1x1(-3);
+    uint1x1 u1 = uint1x1(1);
+    uint1x1 u2 = uint1x1(2);
+    uint1x1 u3 = uint1x1(3);
+    uint1x1 um1 = uint1x1((uint)(-1));
+    float1x1 fm0_5 = float1x1(-0.5);
+    float1x1 f0_5 = float1x1(0.5);
+    float1x1 f1 = float1x1(1);
+    float1x1 f1_5 = float1x1(1.5);
+    float1x1 f2 = float1x1(2);
+
+    // Unary operators, except pre/post inc/dec
+    // CHECK: i32 3, i32 undef
+    Append(output_i, +i3); // Plus
+    // CHECK: i32 -3, i32 undef
+    Append(output_i, -i3); // Minus
+    // CHECK: i32 -4, i32 undef
+    Append(output_i, ~i3); // Not
+    // CHECK: i32 0, i32 undef
+    Append(output_b, !i3); // LNot
```

Comment:
curious what do you get if you apply '!' on float NaN here?

Reply:
it's implemented as != 0, so presumably you get false.